m3/src/reporter_impl.cc

```diff
@@ -0,0 +1,438 @@
+      new thrift::M3Client(std::make_shared<TCompactProtocol>(transport_)));
+
+  thread_ = std::thread(&Reporter::Impl::Run, this);
+}
+
+Reporter::Impl::~Impl() {
+  // Signal the background thread to shut down.
+  {
+    std::lock_guard<std::mutex> lock(run_mutex_);
+    run_ = false;
+  }
+
+  // Wait for the background thread to finish.
+  cv_.notify_one();
+  thread_.join();
+
+  transport_->close();
+}
+
+std::unique_ptr<tally::Capabilities> Reporter::Impl::Capabilities() {
+  return std::unique_ptr<tally::Capabilities>(new tally::CapableOf(true, true));
+}
+
+void Reporter::Impl::ReportCounter(
+    const std::string &name, const std::map<std::string, std::string> &tags,
+    int64_t value) {
+  auto counter = CreateCounter(value);
+  auto metric_tags = ConvertTags(tags);
+  ReportMetric(name, metric_tags, counter);
+}
+
+void Reporter::Impl::ReportGauge(const std::string &name,
+                                 const std::map<std::string, std::string> &tags,
+                                 double value) {
+  auto gauge = CreateGauge(value);
+  auto metric_tags = ConvertTags(tags);
+  ReportMetric(name, metric_tags, gauge);
+}
+
+void Reporter::Impl::ReportTimer(const std::string &name,
+                                 const std::map<std::string, std::string> &tags,
+                                 std::chrono::nanoseconds value) {
+  auto timer = CreateTimer(value);
+  auto metric_tags = ConvertTags(tags);
+  ReportMetric(name, metric_tags, timer);
+}
+
+void Reporter::Impl::ReportHistogramValueSamples(
+    const std::string &name, const std::map<std::string, std::string> &tags,
+    uint64_t bucket_id, uint64_t num_buckets, double buckets_lower_bound,
+    double buckets_upper_bound, uint64_t samples) {
+  auto counter = CreateCounter(samples);
+  auto metric_tags = ConvertTags(tags);
+
+  // Add tag for bucket.
+  std::stringstream bucket_stream;
+  bucket_stream << boost::format("%s-%s") %
+                       ValueBucketString(buckets_lower_bound) %
+                       ValueBucketString(buckets_upper_bound);
+
+  thrift::MetricTag bucket_tag;
+  bucket_tag.__set_tagName(HISTOGRAM_BUCKET_NAME);
+  bucket_tag.__set_tagValue(bucket_stream.str());
+  metric_tags.insert(bucket_tag);
+
+  // Add tag for bucket ID.
+  thrift::MetricTag id_tag;
+  id_tag.__set_tagName(HISTOGRAM_BUCKET_ID_NAME);
+  id_tag.__set_tagValue(BucketID(bucket_id, num_buckets));
+  metric_tags.insert(id_tag);
+
+  ReportMetric(name, metric_tags, counter);
+}
+
+void Reporter::Impl::ReportHistogramDurationSamples(
+    const std::string &name, const std::map<std::string, std::string> &tags,
+    uint64_t bucket_id, uint64_t num_buckets,
+    std::chrono::nanoseconds buckets_lower_bound,
+    std::chrono::nanoseconds buckets_upper_bound, uint64_t samples) {
+  auto counter = CreateCounter(samples);
+  auto metric_tags = ConvertTags(tags);
+
+  // Add tag for bucket.
+  std::stringstream bucket_stream;
+  bucket_stream << boost::format("%s-%s") %
```

Comment:
could you cache this string on the histogram type to avoid having to create is each time the Report loop runs?

Reply:
I looked at doing such a thing originally, but it would require us to change the API here to be different from Go and Java, we would need to take a string instead of the bounds directly. Not sure if the added performance is worth having to changing the API.

Reply:
Most users aren't going to run into issues if these differ because they'd only interact with the reporter ctor, and the Scope directly. So I'm more OK with changing the reporter APIs if there's a perf difference, what do you reckon?

Reply:
My main concern would be that it breaks any possibility of using a different backend (e.g. Prometheus) because now you aren't passing the buckets but this m3-specific string representation of them. Not sure if we should make that tradeoff, I can see both ways.

Reply:
If we wanted to avoid this creation each time we can used the CachedStatsReporter interface instead of just the StatsReporter.  What are your thoughts on StatsReporter vs CachedStatsReporter and why to go down the StatsReporter route for the client @jeromefroe ?

Reply:
I went with the `StatsReporter` interface because that's the only one supported by the Java client so it seemed like the lowest common denominator. I didn't want to implement two interfaces since it would essentially double the size of the PR because everything would need to be done twice, once for each interface.

Reply:
As a point of comparison, the Prometheus also client makes the metrics stateless and they are [serialized into the appropriate representation](https://github.com/jupp0r/prometheus-cpp/blob/master/core/src/text_serializer.cc#L119) during scraping.

Reply:
My personal feeling is that we should stick with the simpler `StatsReporter` interface until performance becomes a clear issue.

Reply:
Why boost format? This is relatively straightforward without a huge boost dependency.

Reply:
We already depend on boost because of Thrift, so it seemed we might as well take advantage of it.